feat(dashboard): Add thumbnails to dashboard edit draggable chart list

codyml opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
This PR adds thumbnails to the sidebar of draggable charts in the edit dashboard view and modifies the design of the cards with and without thumbnails enabled.  Thumbnails are shown when the `THUMBNAILS` feature flag is enabled.  If no thumbnail has been generated, the same placeholder that's used on the chart list card view is shown.  Other changes include:
- Layout updates for the text and the `Added` badge, when thumbnails are enabled and when they aren't.
- Text is truncated and shows a tooltip when it's too long to fit in the layout.
- The human-readable chart viz type is displayed instead of the machine-readable name.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

<!--- Skip this if not applicable -->
**Before:**

https://user-images.githubusercontent.com/13007381/178516532-8830f90c-24fd-441c-a864-47c74b618df8.mov

**After (thumbnails enabled):**

https://user-images.githubusercontent.com/13007381/178517080-023dd336-7c35-4784-9efb-28c822a1bf08.mov

**After (thumbnails disabled):**

https://user-images.githubusercontent.com/13007381/178517519-89811718-34e0-4a97-85f3-17f040885f25.mov

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

**Note:** Thumbnail generation is currently broken (see #20673).  In addition, there's a bug preventing the chart metadata from being downloaded (see #20684), so even if thumbnails are generated, any items in the draggable chart list with missing metadata won't show thumbnails either.

1. Run Superset with feature flag `THUMBNAILS=False` and ensure that the updated cards look and work as expected when added and not added to the dashboard. 
2. Run Superset with `THUMBNAILS=True` and ensure that thumbnails appear (or at least that their placeholders do) as expected and that the updated cards look and work as expected when added and not added to the dashboard.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [x] Required feature flags: `THUMBNAILS`
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

codyml committed:
Add chart thumbnails to dashboard edit draggable charts.

codyml committed:
Reorganize hierarchy and add tests.

codyml commented:
@kasiazjc here's the PR!

zhaoyongjie commented on superset-frontend/src/dashboard/components/AddSliceCard/index.jsx:
```diff
@@ -111,14 +132,31 @@ function AddSliceCard({
+  thumbnailUrl,
   visType,
 }) {
+  const showThumbnails = useMemo(
+    () => isFeatureEnabled(FeatureFlag.THUMBNAILS),
+    [],
+  );
+
```
the value of `isFeatureEnabled(FeatureFlag.THUMBNAILS)` will not change when the Superset web application started, so the `useMemo` might be useless.

zhaoyongjie commented:
@codyml please rebase master to fix the code coverage issue.

codyml committed:
Incorporate review suggestions.

zhaoyongjie reviewed (approved):
looks great!

kgabryje commented:
/testenv up FEATURE_THUMBNAILS=true

rusackas commented:
I think we need to make some layout adjustments here... I'll ping you on a separate thread to save everyone the notification headaches :)

zhaoyongjie commented:
I can't get chart snapshot on Dashboard edit list, but gets it on Charts List.

https://user-images.githubusercontent.com/2016594/176653301-6c6ccd50-06f4-47aa-bcec-b5de59d7b3eb.mov

codyml commented:
> I can't get chart snapshot on Dashboard edit list, but gets it on Charts List.
> 
>  chart.snapshot.in.Dashboard.mov

Ah thanks, will look into this.

codyml committed:
Update design and add tooltips.

codyml committed:
Fix missing thumbnails.

codyml committed:
Fix tests.

codyml commented:
> I can't get chart snapshot on Dashboard edit list, but gets it on Charts List.
> 
>  chart.snapshot.in.Dashboard.mov

@zhaoyongjie Finally got around to looking into this, and it's another symptom of the missing metadata bug, which I'm going to tackle separately.  TBD if this is going to be the final fix, but if you comment out [this line](https://github.com/apache/superset/blob/8a5f3aca2a01c5da277556de5002f2d12c39dcff/superset-frontend/src/dashboard/actions/sliceEntities.js#L80) you should see all thumbnails and metadata appear.

codyml committed:
Fix moving viz type label.

rusackas commented:
/testenv up FEATURE_THUMBNAILS=true

zhaoyongjie commented:
this PR should wait for merging [this](https://github.com/apache/superset/pull/20673).

geido commented:
/testenv up FEATURE_THUMBNAILS=true

geido reviewed (commented):
Code looks good to me. Now waiting for a test env for manual testing. You might need to rebase to fix that codecov issue.

zhaoyongjie commented:
@codyml the editor bar on the Dashboard cannot preview the thumbnails, but I can preview in the dashboard list.

![image](https://user-images.githubusercontent.com/2016594/180267408-9d1a81b7-9ac1-4395-8a2a-6cf0a61ee108.png)

codyml commented:
> @codyml the editor bar on the Dashboard cannot preview the thumbnails, but I can preview in the dashboard list.
> 
> ![image](https://user-images.githubusercontent.com/2016594/180267408-9d1a81b7-9ac1-4395-8a2a-6cf0a61ee108.png)

@zhaoyongjie Replied about this issue above, I think this is unrelated and fixed by https://github.com/apache/superset/pull/20684, which just got merged yesterday.  Just rebased over it, can or @geido run `/testenv up FEATURE_THUMBNAILS=true` and see if it's fixed?

zhaoyongjie commented on superset-frontend/src/dashboard/components/AddSliceCard/index.jsx:
```diff
@@ -0,0 +1,276 @@
+/**
```
Just noticed this is a JSX, we should migrate to TSX for the new component.

zhaoyongjie reviewed (approved):
awesome! I have got the screenshot in the edit bar on the dashboard. leave a no-blocking suggestion, migrate to TSX in separated PR or this PR.

![image](https://user-images.githubusercontent.com/2016594/180384565-28dc2bad-2af3-4744-82eb-19c6dd4d0c24.png)

codyml commented:
@geido found an issue with metadata still being missing, need to look into it:
<img width="1728" alt="Screen Shot 2022-07-22 at 10 44 59 AM" src="https://user-images.githubusercontent.com/13007381/180464267-5d67836c-33cd-4b4f-bfa5-3698e71bec16.png">

zhaoyongjie commented:
/testenv up FEATURE_THUMBNAILS=true

zhaoyongjie commented:
/testenv up FEATURE_THUMBNAILS=true FEATURE_THUMBNAILS_SQLA_LISTENERS=true

codyml committed:
Convert AddSliceCard to TS, update hierarchy.

zhaoyongjie commented:
/testenv up FEATURE_THUMBNAILS=true FEATURE_THUMBNAILS_SQLA_LISTENERS=true

zhaoyongjie commented:
/testenv up FEATURE_THUMBNAILS=true

codyml commented:
@geido @zhaoyongjie Confirmed that missing metadata in ephemeral environment was due to ephemeral environment being spun up before PR was rebased over https://github.com/apache/superset/pull/20684.

rusackas reviewed (approved):
LGTM
